00X-OFEP-provider-flag-metadata.md

```diff
@@ -0,0 +1,37 @@
+## Surfacing flag metadata
+
+## State: DRAFTING
+This proposal lays out a mechanism for flag providers to surface arbitrary flag metadata to Open Feature, and for hooks to access this metadata.
+
+
+## Background
+
+Flag providers maintain metadata about a feature flag, and there are scenarios where a hook might benefit from being able to access this metadata. For example, a flag provider could surface a `management-url` attribute which an open-telemetry hook could then add to an otel span, allowing someone viewing a trace for a feature-flagged operation to easily navigate to a webpage describing that flag in detail.
+
+Different providers will expose different kinds of metadata, and different Application Integrators will want to consume that metadata in different ways. As such, the goal of this proposal is to provide a simple way for the Open Feature runtime to pass generic metadata from a provider to a hook, without Open Feature itself understanding the semantics of that metadata.
+
+## Non-goals
+
+In the future we may wish to standardize the semantics of some key metadata attributes which are common amongst many providers (e.g. `expiration-date`, `owner`, `management-url`) so that provider-agnostic hooks can be created to consume these standard attributes, but that is explicitly NOT in the scope of this OFEP - we would prefer to see which common attributes naturally emerge and then "pave those cow paths".
+
+## Proposal
+
+The Open Feature spec will be extended in two ways:
+- a flag provider will be able to provide a bag of metadata attributes for a feature flag. This bag will be a set of key-value pairs, where the key will be a string and the value will be a small set of primitive values.
```

Comment:
+1 on the 'may' rather than 'must'. In the simple cases, we don't want to force providers to return an empty bag. The OF SDK should default to an empty bag if the provider does nothing.

Reply:
> I wonder if it is worth making preparations for secure/sensitive and non-secure/benign meta data? Something like an evaluation timestamp might be treated in a different way to a management or internal system URL. My worry would be that sensitive metadata then starts getting logged/surfaced without the knowledge or explicit approval of the developer.
> 
> I've not used OpenTelemetry - does it handle this sort of thing or not?

AFAIK otel doesn't support this feature, and IIRC it's actually causes some security issues with baggage that is accidentally propagated outside of internal otel tooling into 3rd party systems. For example, you call into a SaaS provider's APIs, and you get back a response with baggage describing the internal state of their system 😬 

However, I still feel that this is really more the responsibility of the provider to decide whether it should send sensitive metadata to the client or not. For client-side code anything you put on the wire is going to be available regardless, so seems to me that as a provider you'd be forced to just never put that metadata on the wire.

Reply:
ea68f0b makes this clearer